touch up network configuration

cjdoucette opened the pull request:
The patch **lls: add initialization and cache** creates the basic shape of the LLS block and creates a resolution cache. The patch **lls: add ARP processing and cache** instantiates one of those caches for ARP entries and adds the ability to send/receive ARP packets.

AltraMayor commented on lib/net.c:
```diff
@@ -696,36 +696,39 @@ init_iface(struct gatekeeper_if *iface)
-	iface->id = (uint8_t)ret;
+		iface->id = (uint8_t)ret;
+
+		for (i = 0; i < iface->num_ports; i++) {
```
If `iface->num_ports == 1`, this loop will go once, but there won't be a bond interface. You may want to move the whole loop into the `else` clause above.

cjdoucette replied:
I think the diff is a little hard to read, but the loop is already in the `else` clause here.

AltraMayor commented on lua/gatekeeperc.lua:
```diff
@@ -31,6 +31,11 @@ struct ggu_config {
+struct lls_config {
+	unsigned int lcore_id;
+	int          back_iface_enabled;
+};
```
What about the comment about hidden fields?

cjdoucette replied:
Added.

AltraMayor commented on include/gatekeeper_lls.h:
```diff
@@ -19,6 +19,69 @@
+
+struct lls_config {
+	unsigned int      lcore_id;
+	int               back_iface_enabled;
```
Shouldn't `back_iface_enabled` be part of `struct net_config`?

cjdoucette replied:
Yes, thanks. While considering adding it to `struct net_config`, it occurred to me that our code currently assumes that there is a back interface, which is not the case for Grantor, for example. So I broke this addition of `back_iface_enabled` into a separate commit, so that if a back interface is not present, the code does not try to initialize it. Also, blocks that depend on a back interface will now throw an error.

AltraMayor commented on include/gatekeeper_lls.h:
```diff
@@ -19,6 +19,69 @@
+	struct rte_mempool     *entry_pool;
+
+	/* Mailbox to hold requests for changing the cache. */
+	struct mailbox         mb;
```
I've been thinking of mailboxes being associated with blocks. Why did you choose to associate them with caches?

The way it's now, the more the number of caches, the more mailboxes the LLS block will have to pool.

cjdoucette replied:
I originally chose to associate a different mailbox with each cache because the requests for each cache will be of different lengths because the keys are different lengths, so the mempool would need to give out requests of diffferent sizes.

But instead, I've now made just one mailbox for LLS by having the mempool allocate requests with the maximum key size that will be used.

AltraMayor commented on lls/cache.h:
```diff
@@ -0,0 +1,95 @@
+	uint8_t port_id, int inc_refcnt);
+
+/* Add an LLS entry for the key @ip_key and address @addr on @port_id. */
+int lls_cache_add(struct lls_cache *cache, const void *ip_key,
```
How can a block use `lls_cache_get()`, `lls_cache_refcnt()`, and `lls_cache_add()` in order to express interest on an entry, and avoid having to look it up when it needs it? That is, the block wants to inform LLS of the interest on an entry, and it wants to use the entry directly. Otherwise, there will be tons of LLS lookups to ship packets out.

cjdoucette replied:
This is the only comment that I didn't address in my rebase.

We could have the cache directly return the cache entries to the block so that the block can store them with the IP addresses its interested in. Even if the entry changes, the block will always have a direct reference to it and can use the updated Ethernet address. But this would only work for blocks that communicate with a constant number of IP addressess; otherwise, they would have to do a lookup of their saved entries anyway.

Another option might be to add a reference to the corresponding LLS cache entry for each entry in the LPM table, so that when a block does a lookup to find the next hop IP address, it also gets the associated LL address.

But other than those cases (a constant number of interests or doing two lookups in one), I'm not sure how an entry can always be there without needing a lookup.

cjdoucette replied:
I think the best way to do this is to indeed return a pointer to the actual entry that the block should hold onto, instead of having to call `lls_cache_get()` which does a lookup and fills-in a provided `struct ether_addr`.

Note that since the hash table is implemented using cuckoo hashing, an entry will not always be at the same index of the hash table -- adding another entry could bump a previously-stored entry to another position. Therefore, we need a mempool of entries, and each functional block will retain a reference to the entry itself instead of an index into the hash table or anything else.

This will take a few minor changes, I will rebase and let you know when I'm finished.

mengxiang0811 commented on lls/cache.c:
```diff
@@ -0,0 +1,463 @@
+	}
+
+	req = mb_alloc_entry(&cache->mb);
+	if (req == NULL) {
```
Given there is only one line in the ``if`` statement, we may remove the braces here.

cjdoucette replied:
Removed.

AltraMayor commented on lls/main.c:
```diff
@@ -64,9 +116,24 @@ process_pkts(__attribute__((unused)) struct lls_config *lls_conf,
 		ether_type = eth_hdr->ether_type;
+		/*
+		 * Calculate offset to skip over VLAN header, if present.
+		 * TODO If VLAN stripping is enabled, we can skip this.
```
This TODO is even bigger, we need to add VLAN headers on the interfaces that require it. The GK block isn't doing anything about it.

cjdoucette replied:
I added a separate commit that cleans up some of the old `TODO` tags and adds one for VLANs. The stripping side should involve just a simple configuration change (following the assumption that current hardware supports our needs), but at least our current NIC does not support VLAN insertion, so that will need to be done in software.

mengxiang0811 commented on lls/cache.c:
```diff
@@ -0,0 +1,463 @@
+	}
+
+	req = mb_alloc_entry(&cache->mb);
+	if (req == NULL) {
```
Given there is only one line of code in the ``if`` statement, we may remove the braces here.

cjdoucette replied:
Removed.

AltraMayor commented on lls/main.c:
```diff
@@ -64,9 +116,24 @@ process_pkts(__attribute__((unused)) struct lls_config *lls_conf,
+		 * TODO If VLAN stripping is enabled, we can skip this.
+		 */
+		ether_type = eth_hdr->ether_type;
+		next_hdr = (void *)((uint8_t *)(eth_hdr + 1) +
```
This code is assuming that packets are always well formed. While it's okay to assume that a packet always has an Ethernet header, anything above could be missing. Unless you check the length, you may be accessing invalid memory positions.

cjdoucette replied:
Understood, thanks.

mengxiang0811 commented on lls/cache.c:
```diff
@@ -0,0 +1,463 @@
+}
+
+int
+lls_cache_add(struct lls_cache *cache, const void *ip_key,
```
Given that this function and the above ``lls_cache_refcnt()`` have similar code, can we combine these two functions into one?

cjdoucette replied:
Yes, I think that makes a lot of sense. I merged them into `lls_cache_set()`.

mengxiang0811 commented on lls/arp.c:
```diff
@@ -0,0 +1,215 @@
+
+	mp = lls_conf->net->gatekeeper_pktmbuf_pool[
+		rte_lcore_to_socket_id(lls_conf->lcore_id)];
+	created_pkt = rte_pktmbuf_alloc(mp);
```
It's possible that ``rte_pktmbuf_alloc()`` returns a NULL.

cjdoucette replied:
OK, I added a check.

mengxiang0811 commented on lls/main.c:
```diff
@@ -64,9 +116,24 @@ process_pkts(__attribute__((unused)) struct lls_config *lls_conf,
+					bufs[i], eth_hdr, arp_hdr) == -1)
+				goto free_buf;
+			/* ARP reply was sent. */
+			continue;
```
``continue`` here can stop it from freeing the ``bufs[i]``.

cjdoucette replied:
That's OK -- it should only hit the `continue` if `process_arp()` returned 0, which indicates that the buffer was used to transmit a reply (and will be automatically freed).

cjdoucette commented on lls/arp.c:
```diff
@@ -0,0 +1,215 @@
+
+	if (port_id == lls_conf->net->front.id) {
+		source_ip = lls_conf->net->front.ip4_addr;
+		ether_addr_copy(&lls_conf->net->front.eth_addr,
```
Done.

cjdoucette commented on lls/arp.c:
```diff
@@ -0,0 +1,215 @@
+		memset(&eth_hdr->d_addr, 0xFF, ETHER_ADDR_LEN);
+	else
+		/* Unicast to previously-known entry. */
+		ether_addr_copy(ha, &eth_hdr->d_addr);
```
OK, done.

cjdoucette commented on lls/arp.c:
```diff
@@ -0,0 +1,215 @@
+	arp_hdr->arp_data.arp_tip = rte_cpu_to_be_32(*(const uint32_t *)ip_key);
+
+	ret = rte_eth_tx_burst(port_id, tx_queue, &created_pkt, 1);
+	if (ret < 0)
```
Yes, it should be `<=`, and it should be freed if it wasn't sent. Done.

cjdoucette commented on lls/arp.c:
```diff
@@ -0,0 +1,215 @@
+	 */
+	target_ip.s_addr = arp_hdr->arp_data.arp_tip;
+	is_garp = is_garp_pkt(arp_hdr);
+	if ((iface->ip4_addr.s_addr != target_ip.s_addr) && !is_garp)
```
Changed.

cjdoucette commented on lls/arp.c:
```diff
@@ -0,0 +1,215 @@
+
+		/* Need to transmit reply. */
+		num_tx = rte_eth_tx_burst(iface->id, tx_queue, &buf, 1);
+		if (unlikely(num_tx != 1))
```
Done.

cjdoucette commented on lls/arp.c:
```diff
@@ -0,0 +1,215 @@
+	}
+
+	/* Keep IP address in network order for inet_ntop(). */
+	memcpy(&ipv4_addr, ip_key, cache->key_len);
```
OK, done.

cjdoucette commented on lls/arp.c:
```diff
@@ -0,0 +1,215 @@
+	if (entry == NULL)
+		RTE_LOG(INFO, GATEKEEPER,
+			"%s: unresolved (port %hhu) (stale=%s)\n",
+			ip_str, entry->port_id,
```
OK -- the fix here is actually that we should actually test for `!entry->valid`, since an unresolved entry exists in memory (and has port and stale variables), it just doesn't have a valid Ethernet address.

cjdoucette commented on lls/main.c:
```diff
@@ -35,15 +48,53 @@ get_lls_conf(void)
+		vlan_offset = sizeof(struct vlan_hdr);
+		*proto = vlan_hdr->eth_proto;
+
+		if (rte_cpu_to_be_16(ETHER_TYPE_VLAN) == *proto) {
```
In IEEE 802.1ad (QinQ). Can we assume that there will be at most one VLAN header?

AltraMayor replied:
Gatekeeper assumes that there's at most one VLAN header since it only knows about at most one VLAN ID. So we can drop packets with more than a VLAN header. Issuing a warning is advisable.

If we ever receive a request to support multiple VLAN IDs in Gatekeeper, we'll need to change that test and redesign the configuration of Gatekeeper to couple with the multiple VLAN IDs.

AltraMayor replied:
Once you change the code to reflect that, please add a comment using the reasoning in the previous comment.

cjdoucette replied:
I've taken the VLAN tag handling out of the patch entirely, since I added it as a TODO in an earlier patch. Otherwise, LLS would be the only block handling the VLAN.

So until that TODO is done, the assumption in the code is that there is no VLAN tags at all.

AltraMayor replied:
Fine.

cjdoucette commented on lls/cache.h:
```diff
@@ -0,0 +1,95 @@
+#include "gatekeeper_net.h"
+
+/* Set to 1 to enable dumping the cache after it changes. */
+#define DEBUG_LLS_CACHE		0
```
Good idea -- I added it there.

cjdoucette commented:
I think all concerns should be addressed.

Notably, functional blocks can now get a pointer to a cache entry and keep that reference to always use it without needing to do a lookup. There is a small race condition that comes along with this in `__lls_cache_add()`, but I think it is inconsequential and worth the savings.

A functional block should generally call `arp_hold()` on an entry it wants, and then call `arp_resolve()` to actually get the entry, maybe looping until it is available. It can't be assured of getting the entry when `arp_hold()` is called because there might be multiple blocks calling `arp_hold()` and allocating the same entry at the same time, but only the block whose request is processed first will have its entry actually stored in the table.

AltraMayor commented:
Patch **gatekeeper: review TODO tags** is ready for merge.

AltraMayor commented:
Patch **net: fix spacing of net structs** is ready for merge.

AltraMayor commented:
Patch **net: add ability to turn off back iface** is ready for merge.

AltraMayor commented on lls/cache.c:
```diff
@@ -0,0 +1,428 @@
+	}
+
+	RTE_ASSERT(snprintf(entry_pool_name, sizeof(entry_pool_name),
+		"%s_entry_pool", cache->name < sizeof(entry_pool_name)));
```
This code is wrong. It's funny it compiles. Look for an example in the source of Gatekeeper.

cjdoucette replied:
Wow, that is surprising, but I see that you've raised an issue about whether `RTE_ASSERT` is the right macro to use. For now, I've corrected the syntax and kept `RTE_ASSERT` until we fix that potential problem in all the code.

AltraMayor commented on lls/cache.h:
```diff
@@ -0,0 +1,128 @@
+ * Therefore, it should be set to the maximum key length
+ * needed, but no more to avoid wasting mempool allocation.
+ */
+#define LLS_CACHE_MAX_KEY_LEN	4
```
The unit is byte, isn't it? If so, shouldn't it be 16 bytes = 128 bits?

cjdoucette replied:
Yes, thanks.

AltraMayor commented on lls/cache.h:
```diff
@@ -0,0 +1,128 @@
+	 * NOTE
+	 *	This is a variable-length field.
+	 */
+	const void             *ip_key[0];
```
Given that you have `LLS_CACHE_MAX_KEY_LEN`, do you still need this variable length field?

cjdoucette replied:
Removed.

AltraMayor commented on lls/cache.h:
```diff
@@ -0,0 +1,128 @@
+	 *       requests are not answered but there is still
+	 *       interest in this entry.
+	 */
+	int               valid;
```
Is `valid` equal to `refcnt > 0 && !stale`? If so, isn't it better to create an inline function for it?

AltraMayor replied:
Good simplification.

I think that the scanning code should take advantage of unicast transmissions. But instead of having a state field for the entry, we could have a state field to scanning code. Example, when the entry becomes stale (or before it becomes stale), a unicast packet is sent. If the scanning code still finds a stale entry after a unicast transmission has been sent previously, then it issues a broadcast packet.

The motivation for this suggestion is that, in production, many of the entry will be validated by the unicast packets, and avoiding broadcast reduces load on the physical network.

AltraMayor commented on lls/main.c:
```diff
@@ -16,11 +16,203 @@
+			struct ether_hdr *);
+
+		if (!is_broadcast_ether_addr(&eth_hdr->d_addr) &&
+		    !is_same_ether_addr(&eth_hdr->d_addr, &iface->eth_addr))
```
When bonding is being used, do all slave interfaces assume the MAC address of the bond?

cjdoucette replied:
In the types of bonding that we're using, yes: http://dpdk.org/doc/guides-16.07/prog_guide/link_bonding_poll_mode_drv_lib.html#configuration

AltraMayor replied:
Adding that comment and link explains why there's no bug in that test.

AltraMayor commented:
I don't have time to go over patch **lls: add ARP processing and cache** today, but I have a comment for it: ARP should only be initialized for a given interface (i.e. front or back) if `GK_CONFIGURED_IPV4` is in field `configured_proto` of `struct gatekeeper_if` of the interface.

cjdoucette commented on lls/cache.c:
```diff
@@ -0,0 +1,428 @@
+			cache->name);
+		rte_mempool_put(cache->entry_pool, new_entry);
+	} else {
+		struct lls_cache_entry *existing_entry = cache->data[ret];
```
In previous reviews, we removed some `RTE_ASSERT` because we wanted to fail gracefully in code after setup is completed. Following that, should we just add an error message here instead?

cjdoucette replied:
For now, I added these in.

AltraMayor replied:
There are errors that we can and should handle to avoid bringing a server down. Some examples of these cases are malformed, malicious, and unexpected packets, and lack of memory to do a given task. However, there are conditions that should NEVER happen unless there is a bug in Gatekeeper. When a bug happens, the best we can do is to catch it as soon as possible to fix it.

According to the documentation of DPDK, the assert I suggested above should never happen.

cjdoucette commented on lls/cache.c:
```diff
@@ -0,0 +1,428 @@
+{
+	int ret;
+
+	if (unlikely(cache == NULL)) {
```
There's nothing that prevents or discourages other blocks from calling functions like `arp_put()`, `arp_hold()`, etc. until after the LLS block is initialized (and these functions try to use the cache). I mean, blocks *should* wait to do this until after the network starts and all blocks are running, but there's nothing enforcing it.

cjdoucette commented on lls/cache.c:
```diff
@@ -0,0 +1,428 @@
+			lls_entry->stale = true;
+			if (lls_entry->refcnt > 0) {
+				/* Unicast request. */
+				cache->xmit_req(lls_entry->port_id, ip_key,
```
If the IP address now maps to another MAC address, the unicast should not get a response. When the scan happens again, if a response has still not been heard the broadcast will happen.

I think it would also be OK to remove the unicast and just always do broadcast, since entries will only turn stale every two hours or so.

AltraMayor replied:
I like the idea of having a unicast here. The cost is very low, and it may save the entry. But your explanation for why there is no bug here is also important.

cjdoucette commented on lls/cache.c:
```diff
@@ -0,0 +1,428 @@
+ * Size of mempool cache for LLS cache entries.
+ * This should be a divisor of LLS_CACHE_ENTRIES.
+ */
+#define LLS_ENTRIES_MP_CACHE_SIZE	77
```
Agreed -- dropped.

cjdoucette commented on lls/cache.c:
```diff
@@ -0,0 +1,428 @@
+{
+	int32_t ret = rte_hash_add_key(cache->hash, ip_key);
+	if (unlikely(ret == -EINVAL)) {
+		RTE_LOG(ERR, HASH, "Invalid params, %s entry not added\n",
```
It's a tricky fix because we need to be able to convert keys to IP strings for both IPv4 and IPv6, depending on the cache. So I added another function pointer for the cache to accommodate this.

cjdoucette commented on lls/cache.c:
```diff
@@ -0,0 +1,428 @@
+			/* Add placeholder entry and try to resolve it. */
+			__lls_cache_add(cache, ip_key, entry);
+			cache->xmit_req(lls_conf->net->front.id, ip_key, NULL);
+			cache->xmit_req(lls_conf->net->back.id, ip_key, NULL);
```
Deciding whether or not we should send a request on a particular interface gets tricky among these three things:

- Making sure ARP is enabled on the interface.
- Making sure the interface is enabled itself (in the case of the back interface).
- Making sure we're only sending out on an interface that matches the network mask.

I've centralized all of this into one function now (`send_request()` in lls/cache.c), and added a `TODO` tag for the network mask part.

cjdoucette commented on lls/cache.c:
```diff
@@ -0,0 +1,428 @@
+		} else
+			cache->data[ret]->refcnt--;
+
+		if (get_lls_conf()->debug)
```
Thanks, used.

cjdoucette commented:
I added the ability to only initialize ARP on a per-interface basis. This boils down to a few entry points: only initializing the ARP cache and allowing requests to it if at _least_ one of the interfaces has an IPv4 address; only putting the EtherType filter for ARP packets on interfaces that have an IPv4 address; and only sending requests from an interface that has an IPv4 address.

Since we'll also have to do a similar test for ND, I added a function pointer in the cache struct that checks whether that kind of cache is enabled on a given interface.

AltraMayor commented on lls/main.c:
```diff
@@ -16,11 +16,203 @@
+	}
+
+	ret = init_mailbox("lls", MAILBOX_MAX_ENTRIES,
+		sizeof(struct lls_cache_req) + LLS_CACHE_MAX_KEY_LEN,
```
You forgot to drop `LLS_CACHE_MAX_KEY_LEN` here after you changed field `ip_key` of `struct lls_cache_req`.

cjdoucette commented:
While I restructure the LLS block, these three patches touch up the network configuration.
